opt: internal error: selectivity is NaN

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5753958?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5753958?buildTab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on master @ [e9ee21860458d997a8155734dc608cfcd050ef24](https://github.com/cockroachdb/cockroach/commits/e9ee21860458d997a8155734dc608cfcd050ef24):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:265,sqlsmith.go:325,test_runner.go:896: error: pq: internal error: selectivity is NaN
		stmt:
		WITH
			with_59186 (col_324238, col_324239, col_324240)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(
									NULL,
									st_expand(st_expand('BOX(0.18367067277232127 -0.2262819669196806,0.3851655442302596 1.1190807834803405)':::BOX2D::BOX2D, (-1.3721122705826665):::FLOAT8::FLOAT8, (-0.9587580977138188):::FLOAT8::FLOAT8)::BOX2D::BOX2D, NULL::FLOAT8)::BOX2D,
									0.9342913627624512:::FLOAT8
								),
								(
									16955:::INT8,
									'BOX(-0.015460572281356 0.16971978060258852,0.06732289197362273 1.2312981570592125)':::BOX2D,
									0.0:::FLOAT8
								),
								(
									7779:::INT8,
									'BOX(0.1301889493883061 -0.9896844710550265,0.3008170602303508 -0.20219048578850046)':::BOX2D,
									NULL
								)
						)
							AS tab_144162 (col_324238, col_324239, col_324240)
				)
		SELECT
			NULL AS col_324241,
			st_extent('BOX(-1.496378353981207 1.1611097503532846,-0.41882428111056025 2.405317193222179)':::BOX2D::GEOMETRY)::BOX2D
				AS col_324242
		FROM
			defaultdb.public.table1@table1_col1_0_col1_2_key AS tab_144163
			JOIN defaultdb.public.table1@table1_col1_0_col1_2_key AS tab_144164
				JOIN defaultdb.public.table1@table1_col1_0_col1_2_key AS tab_144165 ON
						(tab_144164.col1_6) = (tab_144165.col1_0)
						AND (tab_144164.col1_5) = (tab_144165.col1_5)
						AND (tab_144164.col1_4) = (tab_144165.col1_4) ON (tab_144163.tableoid) = (tab_144164.col1_1)
			RIGHT JOIN defaultdb.public.table1@table1_col1_0_col1_2_key AS tab_144166 ON
					isnan(tab_144163.crdb_internal_mvcc_timestamp::DECIMAL)::BOOL,
			defaultdb.public.table1@[0] AS tab_144167,
			defaultdb.public.table1@[0] AS tab_144168
		GROUP BY
			tab_144163.col1_4, tab_144166.tableoid, tab_144165.col1_0, tab_144165.col1_1;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17683

--- a/pkg/sql/opt/memo/statistics_builder.go
+++ b/pkg/sql/opt/memo/statistics_builder.go
@@ -976,7 +976,7 @@ func (sb *statisticsBuilder) colStatSelect(
 	// filter conditions were pushed down into the input after s.Selectivity
 	// was calculated. For example, an index scan or index join created during
 	// exploration could absorb some of the filter conditions.
-	selectivity := props.MakeSelectivity(s.RowCount / inputStats.RowCount)
+	selectivity := props.MakeSelectivityFromFraction(s.RowCount, inputStats.RowCount)
 	colStat.ApplySelectivity(selectivity, inputStats.RowCount)
 	if colSet.Intersects(relProps.NotNullCols) {
 		colStat.NullCount = 0

--- a/pkg/sql/opt/memo/testdata/stats/select
+++ b/pkg/sql/opt/memo/testdata/stats/select
@@ -3232,3 +3232,71 @@ select
  │                     <--- 0 ------- 100000000000
  └── filters
       └── x:1 = 10 [type=bool, outer=(1), constraints=(/1: [/10 - /10]; tight), fd=()-->(1)]
+
+# Regression test for #84478. Avoid error due to selectivity is NaN.
+exec-ddl
+CREATE TABLE t84478 (
+  col0 FLOAT4,
+  col1 REGPROC NULL,
+  col2 BOX2D NOT NULL,
+  col3 TIMESTAMP NULL,
+  col4 FLOAT8 AS (col0 + NULL) VIRTUAL,
+  UNIQUE (col0) STORING (col1, col2, col3)
+)
+----
+
+norm
+SELECT
+  NULL, 1
+FROM
+  t84478 AS t1
+  JOIN t84478 AS t2
+    JOIN t84478 AS t3 ON
+        (t2.col4) = (t3.col0)
+        AND (t2.col3) = (t3.col3)
+        AND (t2.col2) = (t3.col2) ON (t1.tableoid) = (t2.col1)
+  RIGHT JOIN t84478 AS t4 ON
+      isnan(t1.crdb_internal_mvcc_timestamp::DECIMAL)::BOOL
+GROUP BY
+  t1.col2
+----
+project
+ ├── columns: "?column?":33(unknown) "?column?":34(int!null)
+ ├── immutable
+ ├── stats: [rows=1]
+ ├── fd: ()-->(33,34)
+ ├── distinct-on
+ │    ├── columns: col2:3(box2d)
+ │    ├── grouping columns: col2:3(box2d)
+ │    ├── immutable
+ │    ├── stats: [rows=1, distinct(3)=1, null(3)=1, avgsize(3)=0]
+ │    ├── key: (3)
+ │    └── left-join (cross)
+ │         ├── columns: col2:3(box2d) crdb_internal_mvcc_timestamp:7(decimal)
+ │         ├── multiplicity: left-rows(exactly-one), right-rows(zero-or-more)
+ │         ├── immutable
+ │         ├── stats: [rows=1000, distinct(3)=1, null(3)=1000, avgsize(3)=0]
+ │         ├── scan t84478
+ │         │    ├── computed column expressions
+ │         │    │    └── col4:29
+ │         │    │         └── CAST(NULL AS FLOAT8) [type=float]
+ │         │    └── stats: [rows=1000]
+ │         ├── select
+ │         │    ├── columns: col2:3(box2d!null) crdb_internal_mvcc_timestamp:7(decimal!null)
+ │         │    ├── cardinality: [0 - 0]
+ │         │    ├── immutable
+ │         │    ├── stats: [rows=0, distinct(3)=0, null(3)=0, avgsize(3)=0]
+ │         │    ├── key: ()
+ │         │    ├── fd: ()-->(3,7)
+ │         │    ├── values
+ │         │    │    ├── columns: col2:3(box2d!null) crdb_internal_mvcc_timestamp:7(decimal!null)
+ │         │    │    ├── cardinality: [0 - 0]
+ │         │    │    ├── stats: [rows=0, distinct(3)=0, null(3)=0, avgsize(3)=0]
+ │         │    │    ├── key: ()
+ │         │    │    └── fd: ()-->(3,7)
+ │         │    └── filters
+ │         │         └── isnan(crdb_internal_mvcc_timestamp:7) [type=bool, outer=(7), immutable, constraints=(/7: (/NULL - ])]
+ │         └── filters (true)
+ └── projections
+      ├── NULL [as="?column?":33, type=unknown]
+      └── 1 [as="?column?":34, type=int]

--- a/pkg/sql/opt/props/histogram.go
+++ b/pkg/sql/opt/props/histogram.go
@@ -434,6 +434,9 @@ func (h *Histogram) addBucket(bucket *cat.HistogramBucket, desc bool) {
 // ApplySelectivity reduces the size of each histogram bucket according to
 // the given selectivity, and returns a new histogram with the results.
 func (h *Histogram) ApplySelectivity(selectivity Selectivity) *Histogram {
+	if selectivity == ZeroSelectivity {
+		return nil
+	}
 	res := h.copy()
 	for i := range res.buckets {
 		b := &res.buckets[i]

